cgen: invalid cgen with map access when creating a struct in match

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** 5e1a2f6
**OS:** macos, macOS, 12.5.1, 21G83

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
struct Test {
	a string
}

fn main() {
	a := map[string]string{}
	b := match 'test' {
		'test' {
			Test{
				a['test'] or { '' } // issue is here
			}
		}
		else {
			Test{''} 
		}
	}
	println(b)
}
```


**What did you expect to see?**
It runs.
 
**What did you see instead?**
```c
/tmp/v_501/test.9642852598895935192.tmp.c:2131:4: warning: expression result unused [-Wunused-value]
  (*(int*)_t1.data);
   ^~~~~~~~~~~~~~~
/tmp/v_501/test.9642852598895935192.tmp.c:12030:73: error: expected ')'
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                                        ^
/tmp/v_501/test.9642852598895935192.tmp.c:12030:62: note: to match this '('
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                    ^
/tmp/v_501/test.9642852598895935192.tmp.c:12030:145: error: expected expression
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                                                                                                                ^
/tmp/v_501/test.9642852598895935192.tmp.c:12030:145: error: expected ':'
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                                                                                                                ^
                                                                                                                                                                : 
/tmp/v_501/test.9642852598895935192.tmp.c:12030:60: note: to match this '?'
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                  ^
/tmp/v_501/test.9642852598895935192.tmp.c:12030:145: error: expected expression
        main__Test b = ((string__eq(_SLIT("test"), _SLIT("test")))? (           string* _t2 = (string*)(map_get_check(ADDR(map, a), &(string[]){_SLIT("test")}));
                                                                                                                                                                ^
/tmp/v_501/test.9642852598895935192.tmp.c:12032:7: error: use of undeclared identifier '_t2'; did you mean '_t1'?
                if (_t2) {
                    ^~~
                    _t1
/tmp/v_501/test.9642852598895935192.tmp.c:12031:18: note: '_t1' declared here
                _option_string _t1 = {0};
                               ^
/tmp/v_501/test.9642852598895935192.tmp.c:12033:39: error: use of undeclared identifier '_t2'
                        *((string*)&_t1.data) = *((string*)_t2);
                                                           ^
/tmp/v_501/test.9642852598895935192.tmp.c:12043:43: error: expected ';' after expression
((main__Test){.a = (*(string*)_t1.data),})) : (((main__Test){.a = _SLIT("."),})));
                                          ^
                                          ;
/tmp/v_501/test.9642852598895935192.tmp.c:12043:2: warning: expression result unused [-Wunused-value]
((main__Test){.a = (*(string*)_t1.data),})) : (((main__Test){.a = _SLIT("."),})));
 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/test.9642852598895935192.tmp.c:12043:43: error: expected expression
((main__Test){.a = (*(string*)_t1.data),})) : (((main__Test){.a = _SLIT("."),})));
                                          ^
2 warnings and 8 errors generated.
builder error: 
==================
C error. This should never happen.
```

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -28,7 +28,7 @@ fn (mut g Gen) need_tmp_var_in_match(node ast.MatchExpr) bool {
 			if branch.stmts.len == 1 {
 				if branch.stmts[0] is ast.ExprStmt {
 					stmt := branch.stmts[0] as ast.ExprStmt
-					if stmt.expr in [ast.CallExpr, ast.IfExpr, ast.MatchExpr]
+					if stmt.expr in [ast.CallExpr, ast.IfExpr, ast.MatchExpr, ast.StructInit]
 						|| (stmt.expr is ast.IndexExpr
 						&& (stmt.expr as ast.IndexExpr).or_expr.kind != .absent) {
 						return true

--- a/vlib/v/tests/match_expr_with_struct_init_test.v
+++ b/vlib/v/tests/match_expr_with_struct_init_test.v
@@ -1,0 +1,17 @@
+struct Test {
+	a string
+}
+
+fn test_match_expr_with_struct_init() {
+	a := map[string]string{}
+	b := match 'test' {
+		'test' {
+			Test{a['test'] or { '' }}
+		}
+		else {
+			Test{''}
+		}
+	}
+	println(b)
+	assert b.a == ''
+}

